sql/src/main/java/io/crate/expression/scalar/timestamp/TimezoneFunction.java

```diff
@@ -0,0 +1,128 @@
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.expression.scalar.timestamp;
+
+import io.crate.data.Input;
+import io.crate.expression.scalar.ScalarFunctionModule;
+import io.crate.metadata.BaseFunctionResolver;
+import io.crate.metadata.FunctionIdent;
+import io.crate.metadata.FunctionImplementation;
+import io.crate.metadata.FunctionInfo;
+import io.crate.metadata.Scalar;
+import io.crate.metadata.TransactionContext;
+import io.crate.metadata.functions.params.FuncParams;
+import io.crate.metadata.functions.params.Param;
+import io.crate.types.DataType;
+import io.crate.types.DataTypes;
+import io.crate.types.LongType;
+import io.crate.types.TimestampType;
+import io.crate.types.UndefinedType;
+
+import java.time.DateTimeException;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.Locale;
+
+public class TimezoneFunction extends Scalar<Long, Object> {
+
+    public static final String NAME = "timezone";
+    private static final ZoneId UTC = ZoneId.of("UTC");
+    private static final FuncParams FUNCTION_PARAMS = FuncParams.builder(Param.STRING,
+                                                                         Param.of(DataTypes.TIMESTAMPZ,
+                                                                                  DataTypes.TIMESTAMP,
+                                                                                  DataTypes.LONG)).build();
+
+    public static void register(ScalarFunctionModule module) {
+        module.register(NAME, new BaseFunctionResolver(FUNCTION_PARAMS) {
+            @Override
+            public FunctionImplementation getForTypes(List<DataType> signatureTypes) throws IllegalArgumentException {
+                DataType returnType = null;
+                int timestampParamTypeId = signatureTypes.get(1).id();
+                switch (timestampParamTypeId) {
+                    case LongType.ID:
+                    case TimestampType.ID_WITHOUT_TZ:
+                        returnType = DataTypes.TIMESTAMPZ;
+                        break;
+                    case UndefinedType.ID:
```

Comment:
Why is `undefined` allowed?

Reply:
the type of null when null is typed as an input parameter

Reply:
I think that shouldn't be necessary, also long shouldn't be  part of the Param definition. 

The user provided values will then be cast to the signature's types

Reply:
done, the core is now simpler, thank you for this